Block retrieval fixup

AndrejMitrovic opened the pull request:
This is just an optimization bugfix, but actually it uncovered a whole host of issues. The biggest one is that there are numerous methods in the Ledger which return status codes (`bool` return), but which also **throw**.

That means these exceptions can be accidentally caught and swallowed. I attempted marking some nothrow, but there's a very large chain of calls that needs to be marked nothrow to make this work. So we'll have to do this in another PR.

The commit messages explain in more detail.

AndrejMitrovic committed:
Fix a block retrieval bug

Scenario:

- NetworkClient.getBlocksFrom() retrieves
  a set of blocks [10, 11, 12, 13, 14],
  but the last block #14 was invalid.
- The Ledger was updated to point to block #13.
- This code was problematic:
  if (!onReceivedBlocks(blocks, preimages))
      continue LNextNode;
  block_height += blocks.length;

Then `block_height` would not get updated,
so when we tried reading from the next node.
But the blocks would be rejected by the ledger
because we would try to start with height
`#10` again.
`FullNode.addBlocks()` would fail here.

In the end this shouldn't really be a big
problem because the block retrieval is done
in a periodic timer in `FullNode.startPeriodicCatchup()`,
and in it we use the Ledger's latest block height.

This is merely an optimization fix.

AndrejMitrovic committed:
Make getBlocksFrom() nothrow

Logging is already handled internally,
there is no need for client code to
handle this exception manually.

This also uncovered that we were
swallowing exceptions from the Ledger
routines and not just the client.

Ledger routines should be marked
nothrow in the future.

Geod24 commented:
> The biggest one is that there are numerous methods in the Ledger which return status codes (`bool` return), but which also **throw**.

Yeah, and I'm starting to think the status code approach was a bit too eager. We had way too many places where return codes and exceptions alike were ignored, but at least with Exceptions, we will see it.

AndrejMitrovic commented:
>  but at least with Exceptions, we will see it.

Unless they're swallowed. :)

Some kind of `@nodiscard` would really help. Wasn't it proposed recently?

Geod24 commented on source/agora/node/FullNode.d:
```diff
@@ -360,21 +360,23 @@ public class FullNode : API
+        Height last_height;
         foreach (block; blocks)
         {
             if(!this.ledger.enrollment_manager.addPreimages(preimages))
```
This line also looks super buggy, especially now with preimages being potentially disjoint between cycles.
If we get blocks [18, 19, 20, 21] and pre-images for 21, the EM would reject them ATM.

AndrejMitrovic replied:
Oh then I think we should add a `rollback()` call here.

AndrejMitrovic replied:
Now that I look at this function more it looks really sketchy. `addPreimages` is called multiple times...

AndrejMitrovic replied:
I think we need to rework the catching up phase w.r.t. preimages. I'll file a separate issue.

AndrejMitrovic replied:
Scratch that, I'll just fix the function. I've also noticed that this function will return `false` if we receive a duplicate preimage (same height), which means that if we receive a newer preimage and *then* do a `getBlocksFrom + getPreimages` the `addBlocks` function will terminate early and the blocks would be rejected.

Geod24 replied:
IMO it's still broken ATM, because upon receiving PI for 21, the Ledger (at 17) will reject them and so adding block 18 will not work. When we move from distance to height, we will need to accept PI beyond the enrolled cycle (or adjust them to be at the end of the cycle, but there's no downside in accepting them beyond the cycle within a reasonable margin).

Geod24 commented on source/agora/network/Manager.d:
```diff
@@ -905,18 +905,21 @@ public class NetworkManager
-                block_height += blocks.length;
+                try
+                {
+                    block_height = onReceivedBlocks(blocks, preimages);
+                    if (block_height == 0)
+                        continue LNextNode;
```
The special casing on `0` seems odd. Why not return the number of blocks added ?

AndrejMitrovic replied:
Right I seem to have missed this. Will fix.

AndrejMitrovic replied:
In fact I can get rid of the do/while loop now too, because I made the function nothrow.

AndrejMitrovic committed:
Fix addBlocks() to not reject valid blocks

There was a bug with the way `addBlocks` short-circuited
and returned early.

If any of the preimages were rejected (for *any* reason),
further blocks would be rejected.

However, there are at least two cases where the internal
preimage state could mess with `addBlocks` and cause
it to return early:

A) We already have a duplicate preimage internally,
   from one of the previous receivePreimage() calls.
B) The preimage is for a newer enrollment which is in one of the
   `blocks` which we haven't read from yet. We read the
   blocks sequentally but the preimages were being
   added for every block height. As soon as one block
   was accepted, the second one would be rejected because
   of the short-circuit behavior (duplicate preimage found).

AndrejMitrovic commented:
I think this should fix it now. I think we could also optimize the last commit (maybe in another PR?). We don't need to keep trying to re-add preimages which were successfully added. We could filter them by their enrollment start height.

AndrejMitrovic commented:
@kchulj I am placing a bet of one beer (any beer of your choice) that this will fix https://github.com/bosagora/agora/issues/1552 🍺

kchulj commented:
> @kchulj I am placing a bet of one beer (any beer of your choice) that this will fix #1552 🍺

Related to #1083 ?

AndrejMitrovic commented:
> Related to #1083 ?

Somewhat. But there were multiple bugs found in the catchup code here. The net effect is that blocks would be silently rejected. That might explain why there was nothing interesting in the logs in #1552.

Geod24 commented:
> @kchulj I am placing a bet of one beer (any beer of your choice) that this will fix #1552 🍺

I'll take the bet if Ku doesn't :P
I don't think #1552 is related to catch up, although I would love to be wrong.
